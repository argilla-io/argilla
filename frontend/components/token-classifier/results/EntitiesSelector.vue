<template>
  <span class="entities__selector__container">
    <div class="entities__selector">
      <ul v-if="formattedEntities.length" class="entities__selector__options">
        <li
          class="entities__selector__option suggestion"
          :class="[
            `color_${suggestedEntity.colorId}`,
            activeEntity === -1 ? 'active' : null,
          ]"
          v-if="suggestedEntity"
          @click="selectEntity(suggestedEntity)"
        >
          <span>{{ suggestedEntity.text }}</span>
          <span class="entity__sort-code">[space]</span>
        </li>
        <li
          class="entities__selector__option suggestion"
          :class="[
            `color_${lastSelectedEntity.colorId}`,
            activeEntity === -1 ? 'active' : null,
          ]"
          v-else-if="lastSelectedEntity.text"
          @click="selectEntity(lastSelectedEntity)"
        >
          <span>{{ lastSelectedEntity.text }}</span>
          <span class="entity__sort-code">[space]</span>
        </li>
        <li
          v-for="(entity, index) in formattedEntities"
          tabindex="0"
          :focused="activeEntity === index"
          :key="index"
          class="entities__selector__option"
          :class="[
            `color_${entity.colorId}`,
            activeEntity === index ? 'active' : null,
          ]"
          @click="selectEntity(entity)"
        >
          <span>{{ entity.text }}</span>
          <span class="entity__sort-code"
            >[{{ activeEntity === index ? "enter" : entity.shortCut }}]</span
          >
        </li>
      </ul>
      <div v-else class="entities__selector--empty">
        <svgicon name="error" width="17" height="17" />
        <p>No labels available</p>
      </div>
    </div>
  </span>
</template>

<script>
import "assets/icons/error";
import { mapActions } from "vuex";
export default {
  data: () => ({
    activeEntity: -1,
  }),
  props: {
    dataset: {
      type: Object,
      required: true,
    },
    token: {
      type: Object,
      required: true,
    },
    suggestedLabel: {
      type: String,
    },
    formattedEntities: {
      type: Array,
    },
    showEntitiesSelector: {
      type: Boolean,
      required: true,
    },
  },
  computed: {
    lastSelectedEntity() {
      return this.dataset.lastSelectedEntity;
    },
    suggestedEntity() {
      return this.formattedEntities.find(
        (ent) => ent.text === this.suggestedLabel
      );
    },
  },
  mounted() {
    window.addEventListener("keydown", this.keyDown);
  },
  destroyed() {
    window.removeEventListener("keydown", this.keyDown);
  },
  methods: {
    ...mapActions({
      updateLastSelectedEntity:
        "entities/token_classification/updateLastSelectedEntity",
    }),
    async selectEntity(entityLabel) {
      await this.updateLastSelectedEntity({
        dataset: this.dataset,
        lastSelectedEntity: entityLabel,
      });
      this.token.entity
        ? this.$emit("changeEntityLabel", this.token.entity, entityLabel.text)
        : this.$emit("selectEntity", entityLabel.text);
      this.resetActiveEntity();
    },
    resetActiveEntity() {
      this.activeEntity = -1;
    },
    keyDown(event) {
      const cmd = String.fromCharCode(event.keyCode).toUpperCase();
      if (this.showEntitiesSelector && cmd) {
        const element = document.getElementsByClassName("active");
        event.preventDefault();
        // enter
        if (event.keyCode === 13) {
          if (this.activeEntity !== -1) {
            this.selectEntity(this.formattedEntities[this.activeEntity]);
          }
          //space
        } else if (event.keyCode === 32) {
          if (this.suggestedEntity) {
            this.selectEntity(this.suggestedEntity);
          } else if (this.lastSelectedEntity) {
            this.selectEntity(this.lastSelectedEntity);
          }
          //down
        } else if (
          event.keyCode === 40 &&
          this.activeEntity + 1 < this.formattedEntities.length
        ) {
          this.activeEntity++;
          if (element[0] && element[0].offsetTop >= 90) {
            element[0].parentNode.scrollTop = element[0].offsetTop - 2;
          }
          //up
        } else if (event.keyCode === 38 && this.activeEntity >= 0) {
          this.activeEntity--;
          if (element[0]) {
            element[0].parentNode.scrollTop =
              element[0].offsetTop - element[0].offsetHeight - 8;
          }
        } else {
          const entity = this.formattedEntities.find((t) => t.shortCut === cmd);
          if (entity) {
            this.selectEntity(entity);
          }
        }
      }
    },
  },
};
</script>
<style lang="scss" scoped>
.entities {
  &__selector {
    min-width: 220px;
    background: palette(grey, smooth);
    font-weight: 600;
    padding: 0.8em;
    border-radius: $border-radius;
    &__container {
      @include font-size(14px);
      line-height: 1em;
      display: inline-block;
      white-space: pre-line;
      position: absolute;
      left: -30%;
      top: 2em;
      z-index: 9;
    }
    &__options {
      max-height: 142px;
      overflow-y: scroll;
      padding-left: 0;
      margin: 0;
      overscroll-behavior: contain;
      position: relative;
    }
    &__option {
      display: flex;
      transition: all 0.2s ease;
      padding: 0.5em;
      position: relative;
      cursor: pointer;
      margin-top: 2px;
      margin-bottom: 2px;
      &.suggestion {
        margin-bottom: 0.5em;
      }
      span {
        cursor: pointer !important;
      }
    }
    &--empty {
      display: flex;
      align-items: center;
      .svg-icon {
        margin-right: 0.5em;
      }
      p {
        margin: 0;
        font-weight: 400;
      }
    }
  }
}
.entity {
  &.non-selectable,
  &.non-selectable--show-sort-code {
    cursor: default;
    pointer-events: none;
  }
  &__sort-code {
    margin-left: auto;
    margin-right: 0;
    .non-selectable & {
      display: none;
    }
  }
}
</style>
