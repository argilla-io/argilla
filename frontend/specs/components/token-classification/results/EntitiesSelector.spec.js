import { mount } from "@vue/test-utils";
import EntitiesSelector from "@/components/token-classifier/results/EntitiesSelector";

function mountEntitiesSelector() {
  return mount(EntitiesSelector, {
    propsData: {
      dataset: {
        task: "TextClassification",
        lastSelectedEntity: {
          colorId: 14,
          shortCut: "1",
          text: "Arreglo",
        },
        viewSettings: {
          annotationEnabled: false,
        },
        labels: ["A", "B"],
      },
      formattedEntities: [
        {
          colorId: 14,
          shortCut: "1",
          text: "Arreglo",
        },
        {
          colorId: 1,
          shortCut: "2",
          text: "CARDINAL",
        },
      ],
      token: {
        end: 73,
        entity: {
          end: 73,
          end_token: 15,
          label: "Arreglo",
          start: 57,
          start_token: 13,
        },
        origin: "annotation",
      },
      showEntitiesSelector: true,
      suggestedLabel: "GPE",
    },
  });
}

describe("EntitiesSelector", () => {
  let spy = jest.spyOn(console, "error");
  afterEach(() => spy.mockReset());

  test("renders properly", () => {
    const wrapper = mountEntitiesSelector();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
