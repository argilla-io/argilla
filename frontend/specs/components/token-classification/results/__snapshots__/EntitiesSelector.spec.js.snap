// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntitiesSelector renders properly 1`] = `
<span class="entities__selector__container"><div class="entities__selector"><ul class="entities__selector__options"><li class="entities__selector__option suggestion color_14 active"><span>Arreglo</span> <span class="entity__sort-code">[space]</span></li>
<li tabindex="0" class="entities__selector__option color_14"><span>Arreglo</span> <span class="entity__sort-code">[1]</span></li>
<li tabindex="0" class="entities__selector__option color_1"><span>CARDINAL</span> <span class="entity__sort-code">[2]</span></li>
</ul>
</div>
</span>
`;
